File: packages/python/anip-server/src/anip_server/hashing.py
"""Canonical hashing utilities for ANIP audit entries."""
import hashlib
import json
from typing import Any


def compute_entry_hash(entry: dict[str, Any]) -> str:
    """Compute the canonical hash of an audit entry for hash-chain linking."""
    canonical = json.dumps(
        {k: v for k, v in sorted(entry.items()) if k not in ("signature", "id")},
        separators=(",", ":"),
        sort_keys=True,
    ).encode()
    return f"sha256:{hashlib.sha256(canonical).hexdigest()}"


def canonical_bytes(entry: dict[str, Any]) -> bytes:
    """Return canonical JSON bytes of an audit entry for Merkle leaf hashing."""
    return json.dumps(
        {k: v for k, v in sorted(entry.items()) if k not in ("signature", "id")},
        separators=(",", ":"),
        sort_keys=True,
    ).encode()

File: packages/typescript/server/src/hashing.ts
import { createHash } from "crypto";

export function computeEntryHash(entry: Record<string, unknown>): string {
  const canonical = canonicalBytes(entry);
  const hash = createHash("sha256").update(canonical).digest("hex");
  return `sha256:${hash}`;
}

export function canonicalBytes(entry: Record<string, unknown>): string {
  const filtered: Record<string, unknown> = {};
  for (const key of Object.keys(entry).sort()) {
    if (key !== "signature" && key !== "id") {
      filtered[key] = entry[key];
    }
  }
  return JSON.stringify(filtered);
}
